---
title: "HS Rats Round 10.2 Genotyping Summary"
date: "2023-12-19"
output:
  html_document:
    df_print: paged
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is a summary of genotyping results to accompany the 'Round 10.2' Heterogeneous Stock SNP dataset produced by the lab of Dr. Abraham Palmer at UC San Diego ([palmerlab.org](https://www.palmerlab.org)). 

## Data Summary

The dataset comprises 7,358,643 SNP variants for 17,812 individuals. Metadata for these individual samples are available in the accompanying genotyping log ```round10_2_genotype_log.csv```. Metadata include sample IDs and their associated projects, sequencing libraries, sequencing flowcells, read mapping statistics, and quality control designations (pass/fail) for various QC steps. 

## Methods

SNP genotypes were produced using a custom bioinformatic pipeline parallelized to accommodate both double-digest genotyping-by-sequencing data (ddGBS) and low-coverage whole-genome sequencing data (lcWGS). ddGBS libraries were produced following [Gileta et al 2020](https://doi.org/10.1534/g3.120.401325) and lcWGS were produced using the [Twist Bioscience 96-Plex Library Prep Kit](https://www.twistbioscience.com/products/ngs/library-preparation/twist-96-plex-library-preparation-kit?tab=resources). The bioinformatic pipeline for the analysis of sequence data is freely available from GitHub at [https://github.com/Palmer-Lab-UCSD/HS-Rats-Genotyping-Pipeline](https://github.com/Palmer-Lab-UCSD/HS-Rats-Genotyping-Pipeline) (manuscript in prep). Briefly, ddGBS sequencing libraries were demultiplexed using [FastX-Toolkit](http://hannonlab.cshl.edu/fastx_toolkit/commandline.html#fastx_barcode_splitter_usage) and lcWGS libraries were demultiplexed using [fgbio](http://fulcrumgenomics.github.io/fgbio/tools/latest/DemuxFastqs.html). ddGBS sequences were trimmed for quality using [cutadapt](https://cutadapt.readthedocs.io/en/stable/) and lcWGS sequences were trimmed using cutadapt and [bbDuk](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/). All libraries were aligned to the *Rattus norvegicus* [mRatBN7.2](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_015227675.2/) reference genome assembly using [bwa](https://bio-bwa.sourceforge.net/bwa.shtml). SNP genotypes were imputed using [STITCH](https://github.com/rwdavies/STITCH). Imputed genotypes were filtered to keep only those with imputation INFO scores > 0.9. Individual samples missing > 10% of high-INFO genotypes were removed from the dataset, as were samples with heterozygosity rates > 4 or 5 standard deviations from the mean heterozygosity (per library type), and those samples whose phenotypic sex (when available) did not match their genetic sex (determined by read count ratios on the X and Y chromosomes). This dataset contains all samples that passed all filtering and quality control steps. 

## Figures


![**Fig. 1: INFO score distribution.** All imputed SNP loci are high quality, with INFO scores (a measure of imputation quality) > 0.9.](/path/to/round10_2_INFO_0.9_hist.png)

![**Fig. 2: Read mapping per chromosome.** All chromosomes are well-represented with millions of reads mapped to each (across all libraries)](/path/to/round10_2_mapstats_chr_mapped_reads.png)


![**Fig. 3: Percentage of reads mapped per chromosome.** Read depth is well distributed across the genome (across all libraries)](/path/to/round10_2_mapstats_chr_percent_mapped.png)


![**Fig. 4: ddGBS quality control filters.** Round 10.2 ddGBS samples are only those that fall within filtering cutoffs for heterozygosity (orange lines) and missingness (red line)](/path/to/round10_2_sample_qc_het_vs_missing_ddGBS.png)


![**Fig. 5: lcWGS quality control filters.** Round 10.2 lcWGS samples are only those that fall within filtering cutoffs for heterozygosity (orange lines) and missingness (red line)](/path/to/round10_2_sample_qc_het_vs_missing_lcWGS.png)


![**Fig. 6: Sex quality control filters.** All Round 10.2 samples' genetic sex is distinguishable and consistent with phenotypic sex](/path/to/round10_2_postqc_sex_chrX_vs_chrY.png)


![**Fig. 7: GWAS filters.** For conducting GWAS, we recommend conducting further filters on genotypes for minor allele frequency (MAF) and Hardy-Weinberg disequilibrium (following a Hardy-Weinberg equilibrium exact test). Suggested cutoffs of MAF > 0.005 (orange line) and HWE exact test -log(P-value) < 10 (red line) would produce a dataset of 5,620,375 SNPs for association testing. Note: these cutoffs have not already been applied. The set of SNPs displayed here is that which would result by applying the suggested filters on the provided Round 10.2 dataset](/path/to/round10_2_snpstats_INFO_0.9_hwe_vs_MAF0.005.png)
